service/storage_service.cc

```diff
@@ -2368,7 +2383,20 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
                 throw std::runtime_error(message);
             }
 
-            // Step 2: Prepare to sync data
+            // Step 2: Make the node a group 0 non-voter before removing it from the token ring.
+            //
+            // Thanks to this, even if we fail after removing the node from the token ring
+            // but before removing it group 0, group 0's availability won't be reduced.
+            assert(ss._group0);
+            auto raft_id = raft::server_id{host_id.uuid()};
+            bool raft_available = ss._group0->wait_for_raft().get();
+            if (raft_available) {
+                slogger.info("removenode[{}]: making node {} a non-voter in group 0", uuid, raft_id);
+                ss._group0->make_nonvoter(raft_id).get();
+                slogger.info("removenode[{}]: made node {} a non-voter in group 0", uuid, raft_id);
+            }
+
```

Comment:
In decommission, we make the node non-voter after sync data, here in removenode, we do this before sync data. Can we make they behave the same, i.e., making it non-voter after sync data? The step to sync data takes most of the time and has the best chance to fail.

Reply:
I did this on purpose, `removenode` assumes that the node is dead, so it's desirable to make the dead node a non-voter as soon as possible so we improve group 0's availability. There's no upside to keeping it as a voter for the duration of data sync. In `decommission` the situation is different, we're still alive, so it's desirable for us to keep being a voter until the last moment.

Reply:
OK. Another point is that it is not a big deal even if we make the node leaving as non voter in case of failure.